Pipes breaks with ghc-8.8.1 new Monad / MonadFail handling of fail

The new `Monad` instance does not have the method `fail`. That has been moved exclusively to `MonadFail` in `ghc-8.8.1`.

I tried building `pipes` with the alpha release of the `ghc-8.8.1` and the current version results in the following error:

```
    [2 of 6] Compiling Pipes.Core       ( src/Pipes/Core.hs, .stack-work/dist/x86_64-linux/Cabal-2.5.0.0/build/Pipes/Core.o )
    [3 of 6] Compiling Pipes            ( src/Pipes.hs, .stack-work/dist/x86_64-linux/Cabal-2.5.0.0/build/Pipes.o )
    
    /tmp/stack7255/pipes-4.3.9/src/Pipes.hs:426:5: error:
        ‘fail’ is not a (visible) method of class ‘Monad’
        |
    426 |     fail _   = mzero
        |     ^^^^
    
    /tmp/stack7255/pipes-4.3.9/src/Pipes.hs:427:16: error:
        The INLINE pragma for ‘fail’ lacks an accompanying binding
          (The INLINE pragma must be given where ‘fail’ is declared)
        |
    427 |     {-# INLINE fail #-}
        |                ^^^^

```

It should be simple enough to fix with some use of the C pre-processor.

--- a/pipes.cabal
+++ b/pipes.cabal
@@ -54,6 +54,9 @@ Library
         void         >= 0.4     && < 0.8 ,
         semigroups   >= 0.17    && < 0.20
 
+    if impl(ghc < 8.0)
+        Build-depends: fail == 4.9.*
+
     Exposed-Modules:
         Pipes,
         Pipes.Core,

--- a/src/Pipes.hs
+++ b/src/Pipes.hs
@@ -429,7 +429,7 @@ instance Monad m => Monad (ListT m) where
 #if !MIN_VERSION_base(4,13,0)
     fail _   = mzero
     {-# INLINE fail #-}
-#if
+#endif
 
 instance Monad m => MonadFail (ListT m) where
     fail _ = mzero

--- a/src/Pipes.hs
+++ b/src/Pipes.hs
@@ -66,6 +66,9 @@ module Pipes (
 import Control.Monad (void, MonadPlus(mzero, mplus))
 import Control.Monad.Catch (MonadThrow(..), MonadCatch(..))
 import Control.Monad.Except (MonadError(..))
+#if !MIN_VERSION_base(4,9,0)
+import Control.Monad.Fail (MonadFail(..))
+#endif
 import Control.Monad.IO.Class (MonadIO(liftIO))
 import Control.Monad.Reader (MonadReader(..))
 import Control.Monad.State (MonadState(..))
@@ -423,7 +426,13 @@ instance Monad m => Monad (ListT m) where
     {-# INLINE return #-}
     m >>= f  = Select (for (enumerate m) (\a -> enumerate (f a)))
     {-# INLINE (>>=) #-}
+#if !MIN_VERSION_base(4,13,0)
     fail _   = mzero
+    {-# INLINE fail #-}
+#if
+
+instance Monad m => MonadFail (ListT m) where
+    fail _ = mzero
     {-# INLINE fail #-}
 
 instance Foldable m => Foldable (ListT m) where

--- a/src/Pipes.hs
+++ b/src/Pipes.hs
@@ -66,9 +66,7 @@ module Pipes (
 import Control.Monad (void, MonadPlus(mzero, mplus))
 import Control.Monad.Catch (MonadThrow(..), MonadCatch(..))
 import Control.Monad.Except (MonadError(..))
-#if !MIN_VERSION_base(4,9,0)
 import Control.Monad.Fail (MonadFail(..))
-#endif
 import Control.Monad.IO.Class (MonadIO(liftIO))
 import Control.Monad.Reader (MonadReader(..))
 import Control.Monad.State (MonadState(..))

